### ISSUE TITLE
Switch to using unsquashfs for squashfs-based installs in unpackfs

### ISSUE THREAD

**Issue by tsimonq2** [calamares/calamares#1118]
Right now, the way Calamares does installs (regardless of the source type), is by mounting the source as a loop device and using rsync to copy it.

If the source is a squashfs (which it very likely is for most distributions), we should switch to using unsquashfs rather than mounting it as a loop device and then using rsync. The rationale for doing this is that the operation occurs in about half the time. I found this out while playing with [distinst](https://github.com/pop-os/distinst); they have extremely quick install times.

I tested this on two identical VMs, both with 8 GB of RAM and 4 CPU cores (and did not do them concurrently), with the latest Lubuntu daily. I found that walking through the rsync procedure that Calamares follows, just the rsync command, takes **3 minutes and 6 seconds**. Calling unsquashfs with a fresh hard drive image, [using the flags distinst does](https://github.com/pop-os/distinst/blob/d583a016e998b07cf368d8e41a624953731c52e2/crates/squashfs/src/lib.rs#L156), my extraction time was down to **1 minute and 27 seconds**. That's just about half the time, and this is on several high-powered VMs. I wonder what the performance increase would look like on much lower-spec machines.

I understand the case of other filesystems than squashfs, and I still think we should use rsync for those, but would there be any objections to switching to unsquashfs for squashfs images?

**Comment by abucodonosor** [comment]
If your image is type squashfs we use unsquashfs ( if found ) already.

https://github.com/calamares/calamares/blob/master/src/modules/unpackfs/main.py#L204-L224

But if you mean to just use unsquashfs to replace whole rsync process that 
is not going to work., unsquashfs cannot exclude dirs and files..

However if you think unsquashfs can do the job to replace rsync please write an 
replacement/demo modules for unpackfs and show us.

Thx

**Comment by tsimonq2** [comment]
> If your image is type squashfs we use unsquashfs ( if found ) already.
> 
> https://github.com/calamares/calamares/blob/master/src/modules/unpackfs/main.py#L204-L224

You're misunderstanding what that code does.

Perhaps you should read the unsquashfs manpage:
```
       -l, -ls
           list filesystem, but don't unsquash.
```

> But if you mean to just use unsquashfs to replace whole rsync process that
> is not going to work., unsquashfs cannot exclude dirs and files..

Oh, that's true. I see one spot that uses that bit, and perhaps that could be reworked in this case.

> However if you think unsquashfs can do the job to replace rsync please write an
> replacement/demo modules for unpackfs and show us.

I plan on it.

**Comment by abucodonosor** [comment]
To be clear I do not care about that at all.. 
I don't care about unsquashfs options what so ever..

I personally do not use this type installations , I consider 'rsync / copy over / just 
unpack something'  sort broken.

I am more for correctness over speed , is after all an installer.

> But if you mean to just use unsquashfs to replace whole rsync process that
is not going to work., unsquashfs cannot exclude dirs and files..

>Oh, that's true. I see one spot that uses that bit, and perhaps that could be reworked in this case.

I don't think so .. but ofc you can after you unpack all fire rm -rf on these if you
maintain the same exclude list like in the unpackfs modules.

**Comment by adriaandegroot** [comment]
There is also the newer `rawfs` module which might fit this use-case. The exclusion part of the code is there to restrict the rsync to one filesystem, as far as I can see, so unsquashing direct from an unmounted FS would get that for free (unless this is also about not writing files onto other mounted filesystems, in which case I'd say the source FS is a bit silly).

**Comment by jstnlth** [comment]
I was unable to replicate Simon's results consistently. Sometimes rsync ran with the fastest times that unsquashfs did. One BIG difference though is that unsquashfs used nearly 600MB of ram, while rsync barely registers.

Although inconclusive again, it did seemed to be faster when running both of the commands without --progress (or -info). Since the unpack only seems to affect about 5-10% of the overall progress bar anyway, it might be good to just take out all code that incrementally tries to update the progress bar during the unpack.

**Comment by kkofler** [comment]
I don't think that that would be a great idea. The unpacking is easily the slowest step of the whole installation. Getting no feedback at all would make it look like the installer hung up. We should rather allocate a larger percentage of the total installation to `unpackfs`.

**Comment by jstnlth** [comment]
> We should rather allocate a larger percentage of the total installation to `unpackfs`.
I would definitely agree with that.  I would say at least 50%, and probably 70% would be more accurate.

**Comment by adriaandegroot** [comment]
Moving on from this milestone: there's a big rewrite-unpackfs branch to be made, but this release isn't it. The code has been tidied some, so the design is clear(er) to me now.

Some of the things mentioned in comments:
 - this module has weight 12, over other modules which weigh 1. That gives it a bigger part of overall progress, although exactly how much depends on the number of modules alongside. There's another feature request somewhere about configurable job-weights. (Concrete example: this module represents 25-45% progress overall on a Manjaro install)
 - Progress reporting is now much more detailed, with file-counts displayed.

**Comment by adriaandegroot** [comment]
- Configurable job-weights were introduced in 3.2.31.
- The *unpackfsc* module, available from *calamares-extensions* 1.2.0, uses unsquash directly, rather than rsync. So it's a module-choice (config file is very very similar) now.
